File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;
class Solution {
    public int[] solution(String[] gems) {
        int n = gems.length;
        Set<String> gemTypes = new HashSet<>(Arrays.asList(gems));
        int m = gemTypes.size(); //보석 종류 개수 

        Map<String, Integer> findGem = new HashMap<>();
        int start = 0, end = 0;
        int minLen = Integer.MAX_VALUE;
        int answerStart = 0, answerEnd = 0;
        
        while(start < n) {
            while(end<n && findGem.size()<m) {
                findGem.put(gems[end], findGem.getOrDefault(gems[end], 0) + 1);
                end += 1;
            }
            
            if (findGem.size() == m) {
                if (end-start < minLen) {
                    minLen = end-start;
                    answerStart = start+1;
                    answerEnd = end;
                }
            }
            
            findGem.put(gems[start], findGem.get(gems[start])-1);
            if (findGem.get(gems[start]) == 0) {
                findGem.remove(gems[start]);
            }
            start += 1;
        }
        return new int[]{answerStart, answerEnd};
    }
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
from collections import defaultdict
def solution(gems):
    answer = [0, 1e9]
    
    diamond = len(set(gems))
    n = len(gems)
    end = 0
    status = defaultdict(int)
    
    
    for start in range(n):
        while end < n and len(status) < diamond:
            status[gems[end]] += 1
            end += 1
            
        if len(status) == diamond and end-1-start < answer[1] - answer[0]:
            answer = [start+1, end]
        
        status[gems[start]] -= 1
        if status[gems[start]] == 0:
            del status[gems[start]]
        start += 1
            
    return answer